src/DynamoCore/Models/DynamoModelEvents.cs

```diff
@@ -357,6 +358,137 @@ internal virtual void OnComputeModelDeserialized()
+            }
+        }
+
+        /// <summary>
+        /// Event that is fired during the opening of the workspace.
+        ///
+        /// Use the XmlDocument object provided to conduct additional
+        /// workspace opening operations.
+        /// </summary>
+        public event Action<object> WorkspaceOpening;
+        internal void OnWorkspaceOpening(object obj)
+        {
+            var handler = WorkspaceOpening;
+            if (handler != null) handler(obj);
+        }
+
+        /// <summary>
+        /// Occurs when a workspaces is opened
+        /// </summary>
+        public event WorkspaceHandler WorkspaceOpened;
+        internal void OnWorkspaceOpened(WorkspaceModel workspace)
+        {
+            if (WorkspaceOpened != null)
+            {
+                WorkspaceOpened.Invoke(workspace);
+                if (workspace is HomeWorkspaceModel hws)
+                    HandleStorageExtensionsOnWorkspaceOpened(hws);
+            }
+        }
+
+        /// <summary>
+        /// This event is raised right before the shutdown of DynamoModel started.
+        /// When this event is raised, the shutdown is guaranteed to take place
+        /// (i.e. user has had a chance to save the work and decided to proceed
+        /// with shutting down Dynamo). Handlers of this event can still safely
+        /// access the DynamoModel, the WorkspaceModel (along with its contents),
+        /// and the DynamoScheduler.
+        /// </summary>
+        public event DynamoModelHandler ShutdownStarted;
+
+        private void OnShutdownStarted()
+        {
+            if (ShutdownStarted != null)
+                ShutdownStarted(this);
+        }
+
+        /// <summary>
+        /// This event is raised after DynamoModel has been shut down. At this
+        /// point the DynamoModel is no longer valid and access to it should be
+        /// avoided.
+        /// </summary>
+        public event DynamoModelHandler ShutdownCompleted;
+
+        private void OnShutdownCompleted()
+        {
+            if (ShutdownCompleted != null)
+                ShutdownCompleted(this);
+        }
+
+        /// <summary>
+        /// This event is raised when Dynamo is ready for user interaction.
+        /// </summary>
+        public event Action<ReadyParams> DynamoReady;
+        private bool dynamoReady;
+        /// <summary>
+        /// Event that is raised when Dynamo model requests a particular python engine
+        /// to reset. String parameter is engine name.
+        /// </summary>
+        internal static event Action<string> RequestPythonReset;
+        internal void OnRequestPythonReset(string pythonEngine)
+        {
+            //only reset if current workspace is a homeworkspace
+            //can't guarantee which workspace to mark dirty otherwise
+            if (CurrentWorkspace is HomeWorkspaceModel hmwsm)
+            {
+                RequestPythonReset?.Invoke(pythonEngine);
+                ResetEngine(true);
```

Comment:
I think this is probably overkill, but this function is only triggered manually for now.

Reply:
by manually you mean via new button in preferences ?
When/if we move to something more reactive/automated....we will probably want to reset only the nodes that are affected by the .py reloads right ?

Reply:
Yes, that would be the ideal - BUT identifying those nodes without creating a circular reference is kind of a pain - we can use reflection or string matching if necessary.

Reply:
so by re executing all the nodes we are also taking care (refreshing) of potentially dangling objects that might still reference the old python files right ?